ktor-server/ktor-server-plugins/ktor-server-rate-limit/jvmAndNix/src/io/ktor/server/plugins/ratelimit/RateLimitInterceptors.kt

```diff
@@ -46,17 +49,30 @@ private fun PluginBuilder<RateLimitInterceptorsConfig>.rateLimiterPluginBuilder(
             val weight = provider.requestWeight(call, key)
             LOGGER.trace("Using key=$key and weight=$weight for ${call.request.uri}")
 
-            val rateLimiterForCall = registry.computeIfAbsent(provider.name to key) {
+            val providerKey = ProviderKey(provider.name, key)
+            val rateLimiterForCall = registry.computeIfAbsent(providerKey) {
                 provider.rateLimiter(call, key)
             }
 
             val state = rateLimiterForCall.tryConsume(weight)
             provider.modifyResponse(call, state)
-            if (state is RateLimiter.State.Exhausted) {
-                LOGGER.trace("Declining ${call.request.uri} because of to many requests")
-                call.respond(HttpStatusCode.TooManyRequests)
-            } else {
-                LOGGER.trace("Allowing ${call.request.uri}")
+            when (state) {
+                is RateLimiter.State.Exhausted -> {
+                    LOGGER.trace("Declining ${call.request.uri} because of too many requests")
+                    call.respond(HttpStatusCode.TooManyRequests)
+                }
+
+                is RateLimiter.State.Available -> {
+                    clearOnRefillJobs[providerKey]?.cancel()
+                    clearOnRefillJobs[providerKey] = application.launch {
+                        delay(state.refillAtTimeMillis - getTimeMillis())
+                        // there is a race here, where we can remove limiter that was just consumed,
```

Comment:
Is it a rare situation when such race can occur?

Reply:
One coroutine should retrieve the rate limiter, while another resumed after the `delay` above. The effect of it will be that request will not be counted in a limit, allowing more requests to be served than specified. 

I think, comparing these two options, it's better to have the race than locking server, even if there is a lock per key.